File: examples/ros2-pubsub/cpp_pubsub/src/publisher.cpp
// Copyright (c) 2026 WHEELTEC ROS Textbook Contributors
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class CoursePublisher : public rclcpp::Node
{
public:
  CoursePublisher()
  : Node("course_cpp_talker")
  {
    publisher_ = create_publisher<std_msgs::msg::String>("/course/chatter", 10);
    timer_ = create_wall_timer(500ms, std::bind(&CoursePublisher::publish_message, this));
  }

private:
  void publish_message()
  {
    std_msgs::msg::String message;
    message.data = "C++ says hello #" + std::to_string(sequence_++);
    RCLCPP_INFO(get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::size_t sequence_{0};
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CoursePublisher>());
  rclcpp::shutdown();
  return 0;
}

File: examples/ros2-pubsub/py_pubsub/py_pubsub/publisher.py
# Copyright (c) 2026 WHEELTEC ROS Textbook Contributors
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.

"""Publish numbered textbook messages on a fixed ROS 2 topic."""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class CoursePublisher(Node):
    """Publish one message every half-second."""

    def __init__(self) -> None:
        super().__init__("course_py_talker")
        self.publisher = self.create_publisher(String, "/course/chatter", 10)
        self.sequence = 0
        self.timer = self.create_timer(0.5, self.publish_message)

    def publish_message(self) -> None:
        """Build and publish the next numbered message."""
        message = String()
        message.data = f"Python says hello #{self.sequence}"
        self.publisher.publish(message)
        self.get_logger().info(f"Publishing: '{message.data}'")
        self.sequence += 1


def main(args=None) -> None:
    """Run the publisher until interrupted."""
    rclpy.init(args=args)
    node = CoursePublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
